### ISSUE TITLE
Bug: `no-unused-expressions` is inconsistent

### ISSUE THREAD

**Issue by Rudxain** [eslint/eslint#19216]
### Environment

Node version: v20.18.1
npm version: v10.9.2
Local ESLint version: v9.16.0 (Currently used)
Global ESLint version: Not found
Operating System: linux 6.11.10-amd64

### What parser are you using?

Default (Espree)

### What did you do?

```js
/*eslint no-global-assign: "error"*/
/*eslint no-unused-expressions: "error"*/

// state
let mut = false

// no recursion
const B = Boolean
// bypass `no-global-assign`
Object.defineProperty(globalThis, 'Boolean', {
	get() {
		mut = true
		return B
	}
})

// false positive
Boolean
console.log(mut) // "true"
```

### What did you expect to happen?

From [the official docs](https://eslint.org/docs/latest/rules/no-unused-expressions):
> This rule does not apply to function calls or constructor calls with the new operator, because they could have side effects on the state of the program.

If that's the case, then every `writable && configurable` property (on `globalThis`, `prototype`, etc...) should be considered impure, and therefore "never unused"

### What actually happened?

```
  18:1  error  Expected an assignment or function call and instead saw an expression  no-unused-expressions

✖ 1 problem (1 error, 0 warnings)
```

### Link to Minimal Reproducible Example

https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tZ2xvYmFsLWFzc2lnbjogXCJlcnJvclwiKi9cbi8qZXNsaW50IG5vLXVudXNlZC1leHByZXNzaW9uczogXCJlcnJvclwiKi9cblxuLy8gc3RhdGVcbmxldCBtdXQgPSBmYWxzZVxuXG4vLyBubyByZWN1cnNpb25cbmNvbnN0IEIgPSBCb29sZWFuXG4vLyBieXBhc3MgYG5vLWdsb2JhbC1hc3NpZ25gXG5PYmplY3QuZGVmaW5lUHJvcGVydHkoZ2xvYmFsVGhpcywgJ0Jvb2xlYW4nLCB7XG5cdGdldCgpIHtcblx0XHRtdXQgPSB0cnVlXG5cdFx0cmV0dXJuIEJcblx0fVxufSlcblxuLy8gZmFsc2UgcG9zaXRpdmVcbkJvb2xlYW5cbmNvbnNvbGUubG9nKG11dCkgLy8gXCJ0cnVlXCJcbiIsIm9wdGlvbnMiOnsicnVsZXMiOnt9LCJsYW5ndWFnZU9wdGlvbnMiOnsic291cmNlVHlwZSI6InNjcmlwdCIsInBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7fX0sImVjbWFWZXJzaW9uIjoyMDI0fX19

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

That's proof that the rule is unsound, so why not make it stricter by considering the following as "pure" and "unused"?:
```js
Boolean()
new Boolean
new Boolean()
// etc...
```

**Comment by mdjermanovic** [comment]
Hi @Rudxain, thanks for the issue!

> ### What did you expect to happen?
> From [the official docs](https://eslint.org/docs/latest/rules/no-unused-expressions):
> 
> > This rule does not apply to function calls or constructor calls with the new operator, because they could have side effects on the state of the program.
> 
> If that's the case, then every `writable && configurable` property (on `globalThis`, `prototype`, etc...) should be considered impure, and therefore "never unused"

Can you please clarify what did you expect to happen? Is it that the `no-unused-expression` doesn't report on the `Boolean` expression statement?

**Comment by Rudxain** [comment]
`no-unused-expressions` _does trigger_, but as I've shown, the mere presence of `Boolean` caused the `get`ter to be called. I didn't even had to use `Proxy` (weirdly enough, I tried a `Proxy` as my 1st (and 2nd, 3rd, etc...) attempt, but the getter was never called 👀 ...)

The rule assumes that any reference to a [global-obj](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects) is pure, while `new native_pure_constructor` isn't

**Comment by mdjermanovic** [comment]
> The rule assumes that **any reference** (global, local, object, primitive, etc...) is pure, while `new native_pure_constructor` isn't

It's a very edge case that a variable reference triggers a getter, so I think the expected behavior is to report on them. Thus, reporting the `Boolean` expression statement in the original example is the correct behavior. In case it triggers a getter with side effects, an `eslint-disable` comment can be used.
